contracts/zero-ex/contracts/src/external/TransformerDeployer.sol

```diff
@@ -22,64 +22,96 @@ pragma experimental ABIEncoderV2;
         }
         require(deployedAddress != address(0), 'TransformerDeployer/DEPLOY_FAILED');
-        toDeploymentNonce[deployedAddress] = deploymentNonce;
-        emit Deployed(deployedAddress, deploymentNonce, msg.sender);
+        require(isDelegateCallSafe(deployedAddress), 'TransfermDeployer/UNSAFE_CODE');
+        toDeploymentSalt[deployedAddress] = salt;
+        toInitCodeHash[deployedAddress] = keccak256(bytecode);
+        emit Deployed(deployedAddress, salt, msg.sender);
     }
 
-    /// @dev Call `die()` on a contract. Only callable by an authority.
-    /// @param target The target contract to call `die()` on.
-    /// @param ethRecipient The Recipient of any ETH locked in `target`.
-    function kill(IKillable target, address payable ethRecipient)
-        public
-        onlyAuthorized
-    {
-        target.die(ethRecipient);
-        emit Killed(address(target), msg.sender);
+    /// @dev Checks whether a given address is safe to be called via
+    ///      delegatecall. A contract is considered unsafe if it includes any
+    ///      of the following opcodes: CALLCODE, DELEGATECALL, SELFDESTRUCT,
+    ///      SLOAD, and STORE. This code is adapted from
+    ///      https://github.com/dharma-eng/dharma-smart-wallet/blob/master/contracts/helpers/IndestructibleRegistry.sol
+    /// @param target The address to check.
+    /// @return True if the contract is considered safe for delegatecall.
+    function isDelegateCallSafe(address target) public view returns (bool) {
+        uint256 size;
+        assembly { size := extcodesize(target) }
+        require(size > 0, "No code at target.");
+
+        bytes memory extcode = new bytes(size);
+        assembly {
+            extcodecopy(target, add(extcode, 32), 0, size)
+        }
+
+        // Look for any reachable, impermissible opcodes.
+        bool reachable = true;
+        for (uint256 i = 0; i < extcode.length; i++) {
+            uint8 op = uint8(extcode[i]);
+
+            // If the opcode is a PUSH, skip over the push data.
+            if (op > 95 && op < 128) { // pushN
+                i += (op - 95);
+                continue;
+            }
+
+            if (reachable) {
+                // If execution is halted, mark subsequent opcodes unreachable.
+                if (
+                    op == 254 || // invalid
+                    op == 243 || // return
+                    op == 253 || // revert
+                    op == 86  || // jump
+                    op == 0      // stop
+                ) {
+                    reachable = false;
+                    continue;
+                }
+
+                // If opcode is impermissible, contract is unsafe.
+                if (
+                    op == 242 || // callcode
+                    op == 244 || // delegatecall
+                    op == 255 || // selfdestruct
+                    op == 84  || // sload
+                    op == 85     // sstore
```

Comment:
This list looks good. In addition, we probably don't want the Flash Wallet deploying contracts so let's also blacklist `create` and `create2`.

Reply:
Oh, great catch. Anything deployed via `create2` would be implicitly trusted, so it's really important to block that. (`create` is probably less critical but a good idea.)